harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -657,6 +657,55 @@ def test_amp(self, trial_class) -> None:
         )
         controller.run()
 
+    @pytest.mark.skipif(not HAVE_APEX, reason="Apex not available")
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
+    @pytest.mark.gpu
+    def test_apex_checkpointing(self, tmp_path: pathlib.Path) -> None:
+        # There will be num_checkpoints-1 assertions made at the end of this test
+        num_checkpoints = 5
+        training_metrics = {}
+        validation_metrics = {}
+        checkpoint_dir = str(tmp_path.joinpath("apex_checkpoint"))
+        latest_checkpoint = None
+        steps_completed = 0
+
+        def make_workloads(checkpoint_idx: int = 0) -> workload.Stream:
+            # Train as usual
+            trainer = utils.TrainAndValidate()
+            # In order to expect the loss to decrease after loading the first checkpoint,
+            #  make sure the scaler has had enough steps to stop reducing so that training
+            #  can actually proceed.
+            yield from trainer.send(steps=10, validation_freq=10)
+            tm, vm = trainer.result()
+            training_metrics[checkpoint_idx] = tm
+            validation_metrics[checkpoint_idx] = vm
+            # Checkpoint
+            interceptor = workload.WorkloadResponseInterceptor()
+            yield from interceptor.send(workload.checkpoint_workload())
+            nonlocal latest_checkpoint, steps_completed
+            latest_checkpoint = interceptor.metrics_result()["uuid"]
+            steps_completed = trainer.get_steps_completed()
+
+        for c in range(num_checkpoints):
+            controller = utils.make_trial_controller_from_trial_implementation(
+                trial_class=pytorch_onevar_model.OneVarApexTrial,
+                hparams=self.hparams,
+                workloads=make_workloads(c),
+                trial_seed=self.trial_seed,
+                checkpoint_dir=checkpoint_dir,
+                latest_checkpoint=latest_checkpoint,
+                steps_completed=steps_completed,
+                expose_gpus=True,
+            )
+            controller.run()
+
+        for c in range(1, num_checkpoints):
+            prev_checkpoint_last_loss = training_metrics[c - 1][-1]["loss"]
+            this_checkpoint_first_loss = training_metrics[c][0]["loss"]
+            assert (
+                this_checkpoint_first_loss < prev_checkpoint_last_loss
+            ), f"loss did not decrease after checkpoint {c-1}"
```

Comment:
I don't think this is a strong enough assertion.

It is true that the bug we're investigating is about "loss spikes after checkpoint restoring" or something.  But due to the linear model, stable inputs, and simple SGD loss function, I would be surprised to _ever_ see the loss increase, even if we totally messed up saving the scaler state.

Can you instead write the test so that it directly measures that the state of the scaler was properly saved/restored?  I can think of two ways to do it, either:
- direct inspection of scaler state, or
- inferring scaler state by expecting precise loss values, as with the last test you wrote

In either case, you'll need to make sure that the scaler finishes each run in a different state than its initial state, otherwise there's no way to tell if things were preserved or not.

Reply:
Anda had a similar concern in one of our calls, I think. What I can do is check if the scale value changed, and if it did check that there is a good, apparent reason (similar to what we did in the AMP tests).

An obvious problem we'd catch is that the scaler value resets to whatever initial value upon loading the state.

Reply:
Actually, scratch all that. That's just doing the same test as before. What I'm writing now and will push soon is this: the scale value just before saving a checkpoint should be the same as just after loading that checkpoint.

Reply:
that sounds good.  Can you also include an assertion to guarantee that the scaler state change over time?  I.e. that the test isn't passing on accident due to the scaler state always being the initial value.

Reply:
I could, but that'll make the test resemble the AMP autoscaling test.

Reply:
I think avoiding false positives in tests is more important than the DRY principle.

Reply:
We are now using the same assertion code as is being used in the AMP scaling tests.